MAID-2500: Refactor echo address facilities

povilasb opened the pull request:
Update code so that it could be easier reused in Crust.

povilasb committed:
feat/p2p: allow to remove echo address servers only by address

povilasb committed:
feat/echo_addr_server: export RendezvousServerError

Also deduplicate this error between UDP and TCP implementations.
Allows to use this error from the outside of this crate scope.

povilasb committed:
feat/lib: export query_public_addr instead of tcp_query_public_addr

tcp_query_public_addr ended up a lower level function. query_public_addr is easier to use and can be used with both protocols: TCP and UDP.

povilasb committed:
fix/mc: make EncryptedRequest constructor public

This structure is constructed from p2p users (Crust), hence it must be public.

canndrew commented on src/server_set.rs:
```diff
@@ -17,12 +17,15 @@ impl ServerSet {
-        self.iterators.retain(|sender| {
-            sender.unbounded_send((addr.clone(), false)).is_ok()
+    pub fn remove_server(&mut self, addr: SocketAddr) {
+        let server_info = PeerInfo::with_rand_key(addr);
```
What's going on here? Why are we generating a random key and expecting that `PeerInfo` to be present in our `servers` set?

povilasb replied:
Well, first of I modified `ServerSet` to hold `PeerInfo` instead of `SocketAddr`. `PeerInfo` holds both `SocketAddr` and `PublicKey` - that's the required information to securely connect to a peer. Hope this part is clear.

Then, I noticed that in some cases Crust wants to remove STUN servers by IP address - no public key given. So some changes had to be made to `ServerSet`.

`Servers` and `ServerSet` interact with each other via this `UnboundedSender<PeerInfo>`. When we either remove or add servers we send a message to this channel. And in case of removal operation, there's no public key, so I simply generate it randomly.

I guess alternatively, I could make two channels: one for "add server", another one for "remove server".

canndrew replied:
Yeah, having to generate dummy data is a big red flag that the types are wrong somewhere. I'd replace the `(PeerInfo, bool)` in the channel with `enum { Remove(SocketAddr), Add(PeerInfo) }`. And you'll need to decide how to remove the value from the `HashSet`, or replace it with a different data structure, because when you've got there won't remove anything since the public key will be different to what's in the set.

Alternatively, rather than making this change, change the part of crust that wants to remove a server based solely on a `SocketAddr`. This is why I wanted to merge the key into `PaAddr` - if everything is encrypted then a socket address should be completely useless without its key and so it's suspicious that an address is being passed around the code without a key.

povilasb replied:
`PeerInfo` is already hashed only by address https://github.com/ustulation/p2p/pull/56/commits/56bb83f5129733702750c550a190debdaebec678 so no more changes on that are necessary.

Well, in Crust in some places we don't have public key, e.g. https://github.com/maidsafe/crust/blob/dev/src/service.rs#L115 - there's a bootstrap server blacklist which is only IP addresses.

canndrew replied:
> `PeerInfo` is already hashed only by address 56bb83f so no more changes on that are necessary.

I wouldn't expect this to be the case. Is this just a hack to make another hack work?

canndrew replied:
Like, maybe `ServerSet` should hold a `HashMap<SocketAddr, PublicKey>`. Thoughts?

povilasb committed:
feat/hashmap: add remove_random() method

povilasb committed:
refactor/utils: rename hash extensions module

Since HashMap extensions were added, set is not appropriate in module name anymore.
